Yank/experiment.py

```diff
@@ -1179,58 +1303,156 @@ def _validate_solvents(cls, solvents_description):
-                                update_keys={'implicit_solvent': Use(to_openmm_app),
-                                             Optional('nonbonded_method'): Use(to_no_cutoff)},
-                                exclude_keys=['rigid_water'])
-        vacuum_schema = utils.generate_signature_schema(AmberPrmtopFile.createSystem,
-                                update_keys={'nonbonded_method': Use(to_no_cutoff)},
-                                exclude_keys=['rigid_water', 'implicit_solvent'])
-
-        # Common schema to all types of solvents
-        for schema in [explicit_schema, implicit_schema, vacuum_schema]:
-            schema.update({Optional('leap'): cls._LEAP_PARAMETERS_SCHEMA})
-
-        solvent_schema = Schema(Or(explicit_schema, implicit_schema, vacuum_schema))
+        # Create the basic solvent schema, ignoring things which have a dependency
+        # Some keys we manually tweak
+        base_solvent_schema = utils.generate_signature_schema(AmberPrmtopFile.createSystem,
+                                                              exclude_keys=['nonbonded_method'])
+        implicit_solvent_default_schema = {'implicit_solvent': base_solvent_schema.pop('implicit_solvent')}
+        rigid_water_default_schema = {'rigid_water': base_solvent_schema.pop('rigid_water')}
+        nonbonded_cutoff_default_schema = {'nonbonded_cutoff': base_solvent_schema.pop('nonbonded_cutoff')}
+        # Cerberus Schema Processing hierarchy:
+        # Input Value -> {default value} -> default setter -> coerce -> allowed/validate
+        #  Handle the use cases for special keys
+        # nonbonded_method
+        base_solvent_schema['nonbonded_method'] = {
+            'allowed': [value for _, value in mapped_openmm_nonbonded_methods.items()],  # Only use valid openmm methods
+            'coerce': to_openmm_app,  # Cast the string first to valid method
+            'required': True,  # This must be set
+            'default': openmm_nonbonded_strings[0],  # Choose a default mapping
+        }
+        # Explicit solvent keys, populate required and dependencies in batch
+        explicit_only_keys = {
+            'clearance': {
+                'type': 'quantity',
+                'coerce': utils.to_unit_validator(unit.angstrom),
+            },
+            'solvent_model': {
+                'type': 'string',
+                'nullable': True,
+                'validator': is_supported_solvent_model,
+                'default_setter': solvent_model_if_explicit_else_none
+            },
+            'positive_ion': {
+                'type': 'string'
+            },
+            'negative_ion': {
+                'type': 'string'
+            },
+            'ionic_strength': {
+                'type': 'quantity',
+                'coerce': to_unit_validator_unless_none(unit.molar),
+                'default_setter': ionic_strength_if_explicit_else_none,
+                'nullable': True
```

Comment:
It can now be set to null, which behaves the same as not setting it. I simply chose not to document it because we really dont want the users doing that.

As for raising an issue on Cerberus, this is a bit more complicated that that. The problem is the `dependencies` field can only check for literal values, there is no dynamic set. I came up with a way to do that later on in the systems and experiment section for runtime determination, but that won't work here. Here, we would need to set default if and only if the other key has certain values, otherwise the key should not be present. This is different than later on where the key should be set regardless of other keys.

Effectively, we would need to make a `functional_default_dependency` or something where the key depends on another key, and only if that dependency has certain values, otherwise its not set.

Reply:
> the dependencies field can only check for literal values, there is no dynamic set

Sorry if I keep missing something (we can talk in person if it's easier), but why do we need dynamic values here? Wouldn't something like this do the job?
```yaml
ionic_strength:
    type: quantity,
    coerce: to_molar_validator,
    default: 0.0*molar,
    dependencies: {nonbonded_method: [CutoffPeriodic, CutoffNonPeriodic, Ewald, PME]}
```

Reply:
Because the hierarchy of functions it checks (by my own testing) is as follows:
`Input Value -> default -> default setter -> coerce -> allowed/validate/dependencies`

The logic goes like this:
#. If no value is set, then it tries to assign the `default`, 
#. If the field is still empty (i.e. `default` is not set), then it runs through the `default_setter`. 
#. Run the value (if present) through coerce
#. Validate the value against all other rules.

The problem with the `default` and `default_setter` means that the value WILL exist when it checks `dependencies` and `validator`. So in your example, if `nonbonded_method = NoCutoff`, then this would fail because the `ionic_strength` would be set, even if it was set to `None` when it checks against `nonbonded_method`'s value

Reply:
Yeah, this is the reason why I think this could be considered a bug. I don't see a use case where it's useful to evaluate the `dependencies` before the `defaults`.

Reply:
I disagree. I think that the defaults should be set first and then validated against other fields. What is the point of checking the value of another field when the default/setter could change the field after you check it? (e.g. "Do I need to order milk? I had a gallon yesterday, but up to 4 people could have had breakfast. Its probably fine, right?")

I think the formal way to do it would be a type of conditional default, one that triggers only if the dependency field is set to `X`. But based on the dev time of that package, and considering what they already have on their roadmap for patch 1.2, I don't expect this any time soon.

Reply:
Sorry if I wasn't clear, I'm not saying the field should be validated before the default is set. I'm saying that the dependencies should be check before assigning defaults.

Reply:
Same problem. Consider the case if some other field checked if `ionic_strength` was None as a dependency because it only works when `ionic_strength` is its default value, but was not required. In that case, it would check for `ionic_strength`, which has not been set yet because the document's defaults have not been set, does not find `ionic_strength` field, fails, but then the output document says `ionic_strength` is set to the correct value.

Reply:
That could be solved by just processing `ionic_strength` before its dependency. As long as there are no circular dependencies, which are not supported anyway, I don't see a problem with it.

Reply:
I'm not sure I can tell `Cerberus` what order to process the keys in. And in this specific case we have, yes that would solve the problem. But in the general `Cerberus` code, that can't be ensured so I'm saying that even if I did propose such a thing, they would have a very strong argument against doing such a change, one that I would agree with.

Reply:
Oh, maybe I misunderstood. Do you mean that the `default -> ... -> dependencies` is a global order (i.e. first all the defaults are sets to all fields recursively, then they are all coerced, then validated)? In this case, I can see how that would be a problem.